File: 27-08-2025-Length_of_Longest_V-Shaped_Diagonal_Segment.java
import java.util.*;

class Solution {
    int[] dr = {-1, 1, 1, -1};
    int[] dc = {1, 1, -1, -1};

    boolean inBounds(int r, int c, int n, int m) {
        return r >= 0 && r < n && c >= 0 && c < m;
    }

    int dfs(int[][] grid, int r, int c, int dir, boolean canChange, int len, int searchFor) {
        int n = grid.length;
        int m = grid[0].length;
        int maxi = len;

        if (dir != -1) {
            int nr = r + dr[dir];
            int nc = c + dc[dir];
            if (inBounds(nr, nc, n, m) && grid[nr][nc] == searchFor) {
                int nextVal = (searchFor == 2) ? 0 : 2;
                maxi = Math.max(maxi, dfs(grid, nr, nc, dir, canChange, len + 1, nextVal));
            }
        }

        if (dir != -1 && canChange) {
            int ndir = (dir + 1) % 4;
            int nr = r + dr[ndir];
            int nc = c + dc[ndir];
            if (inBounds(nr, nc, n, m) && grid[nr][nc] == searchFor) {
                int nextVal = (searchFor == 2) ? 0 : 2;
                maxi = Math.max(maxi, dfs(grid, nr, nc, ndir, false, len + 1, nextVal));
            }
        }

        if (dir == -1) {
            for (int i = 0; i < 4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                if (inBounds(nr, nc, n, m) && grid[nr][nc] == searchFor) {
                    int nextVal = (searchFor == 2) ? 0 : 2;
                    maxi = Math.max(maxi, dfs(grid, nr, nc, i, canChange, len + 1, nextVal));
                }
            }
        }

        return maxi;
    }

    public int lenOfVDiagonal(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        int maxi = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) {
                    maxi = Math.max(maxi, dfs(grid, i, j, -1, true, 1, 2));
                }
            }
        }
        return maxi;
    }
}

File: 27-08-2025-Length_of_Longest_V-Shaped_Diagonal_Segment.py
class Solution:
    def __init__(self):
        self.dr = [-1, 1, 1, -1]
        self.dc = [1, 1, -1, -1]

    def inBounds(self, r, c, n, m):
        return 0 <= r < n and 0 <= c < m

    def dfs(self, grid, r, c, dir, canChange, length, searchFor):
        n, m = len(grid), len(grid[0])
        maxi = length

        if dir != -1:
            nr, nc = r + self.dr[dir], c + self.dc[dir]
            if self.inBounds(nr, nc, n, m) and grid[nr][nc] == searchFor:
                nextVal = 0 if searchFor == 2 else 2
                maxi = max(maxi, self.dfs(grid, nr, nc, dir, canChange, length + 1, nextVal))

        if dir != -1 and canChange:
            ndir = (dir + 1) % 4
            nr, nc = r + self.dr[ndir], c + self.dc[ndir]
            if self.inBounds(nr, nc, n, m) and grid[nr][nc] == searchFor:
                nextVal = 0 if searchFor == 2 else 2
                maxi = max(maxi, self.dfs(grid, nr, nc, ndir, False, length + 1, nextVal))

        if dir == -1:
            for i in range(4):
                nr, nc = r + self.dr[i], c + self.dc[i]
                if self.inBounds(nr, nc, n, m) and grid[nr][nc] == searchFor:
                    nextVal = 0 if searchFor == 2 else 2
                    maxi = max(maxi, self.dfs(grid, nr, nc, i, canChange, length + 1, nextVal))

        return maxi

    def lenOfVDiagonal(self, grid):
        n, m = len(grid), len(grid[0])
        maxi = 0

        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    maxi = max(maxi, self.dfs(grid, i, j, -1, True, 1, 2))

        return maxi
